split into training and testing data

In [1]:
import pandas as pd
import numpy as np
import sklearn
from sklearn.model_selection import train_test_split

/var/folders/12/nb5wd0692kqdjk5ls7v3ww980000gn/T/ipykernel_20119/1091722933.py:1: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


## MY STREAMING DATA

In [2]:
uncleaned = pd.read_json('/Users/allisonkramp/Desktop/spring24/spotify project!/Spotify Account Data/StreamingHistory_music_0.json')

In [3]:
uncleaned

endTime       artistName                trackName  msPlayed
0     2023-01-24 01:04  Daisy the Great                     Easy    197459
1     2023-01-24 01:09  Daisy the Great                  Routine    300393
2     2023-01-24 02:01  Daisy the Great        Smile Pretty Girl    162242
3     2023-01-24 02:04             Cuco        First Of The Year    170873
4     2023-01-24 02:06     Good Morning                After You    145937
...                ...              ...                      ...       ...
9995  2023-04-25 19:11      beabadoobee         the perfect pair    177533
9996  2023-04-25 19:28       Leith Ross                 Grown Up     46452
9997  2023-04-25 19:28   Peter McPoland           Romeo & Juliet    166137
9998  2023-04-25 19:29            MyKey  Was It Something I Said      2474
9999  2023-04-25 19:32       The Marías            Over the Moon    182900

[10000 rows x 4 columns]

In [4]:
uncleaned_extended = pd.read_json('/Users/allisonkramp/Desktop/spring24/spotify project!/Spotify Extended Streaming History/Streaming_History_Audio_2014-2019_0.json')

In [5]:
uncleaned_extended

ts    username  \
0      2014-06-09T17:53:33Z  allymarie7   
1      2014-06-09T17:53:35Z  allymarie7   
2      2014-06-09T17:53:38Z  allymarie7   
3      2014-06-09T17:53:51Z  allymarie7   
4      2014-06-09T17:57:13Z  allymarie7   
...                     ...         ...   
16787  2019-10-10T14:36:03Z  allymarie7   
16788  2019-10-10T14:40:04Z  allymarie7   
16789  2019-10-10T14:44:38Z  allymarie7   
16790  2019-10-10T14:48:36Z  allymarie7   
16791  2019-10-10T14:52:30Z  allymarie7   

                                             platform  ms_played conn_country  \
0      Android-tablet OS 4.4.2 API 19 (asus, Nexus 7)      11749           US   
1      Android-tablet OS 4.4.2 API 19 (asus, Nexus 7)       2554           US   
2      Android-tablet OS 4.4.2 API 19 (asus, Nexus 7)       3715           US   
3      Android-tablet OS 4.4.2 API 19 (asus, Nexus 7)      14485           US   
4      Android-tablet OS 4.4.2 API 19 (asus, Nexus 7)     202496           US   
...                                               ...        ...          ...   
16787                           iOS 12.2 (iPhone10,1)     211440           US   
16788                           iOS 12.2 (iPhone10,1)     243247           US   
16789                           iOS 12.2 (iPhone10,1)     227744           US   
16790                           iOS 12.2 (iPhone10,1)     240008           US   
16791                           iOS 12.2 (iPhone10,1)     236790           US   

      ip_addr_decrypted user_agent_decrypted  \
0          50.165.42.23              unknown   
1          50.165.42.23              unknown   
2          50.165.42.23              unknown   
3          50.165.42.23              unknown   
4          50.165.42.23              unknown   
...                 ...                  ...   
16787      99.203.15.61              unknown   
16788      99.203.15.61              unknown   
16789      99.203.15.61              unknown   
16790      99.203.15.61              unknown   
16791      99.203.15.61              unknown   

                            master_metadata_track_name  \
0                                    Beauty And A Beat   
1                                 Timber (feat. Ke$ha)   
2      A Little Party Never Killed Nobody (All We Got)   
3                                              Fix You   
4                                 She Looks So Perfect   
...                                                ...   
16787                Pray For Me (with Kendrick Lamar)   
16788                                            Gumbo   
16789                                         TED Talk   
16790                                        Boom (X3)   
16791                                            Anita   

      master_metadata_album_artist_name  \
0                         Justin Bieber   
1                               Pitbull   
2                                Fergie   
3                              Coldplay   
4                   5 Seconds of Summer   
...                                 ...   
16787                        The Weeknd   
16788                          Jay Rock   
16789                          Jonwayne   
16790                    Injury Reserve   
16791                             Smino   

                        master_metadata_album_album_name  ... episode_name  \
0                                       Believe Acoustic  ...         None   
1              Global Warming: Meltdown (Deluxe Version)  ...         None   
2        Music From Baz Luhrmann's Film The Great Gatsby  ...         None   
3                                                    X&Y  ...         None   
4                                   She Looks So Perfect  ...         None   
...                                                  ...  ...          ...   
16787  Black Panther The Album Music From And Inspire...  ...         None   
16788                                              90059  ...         None   
16789                                     

In [6]:
uncleaned_extended.columns

Index(['ts', 'username', 'platform', 'ms_played', 'conn_country',
       'ip_addr_decrypted', 'user_agent_decrypted',
       'master_metadata_track_name', 'master_metadata_album_artist_name',
       'master_metadata_album_album_name', 'spotify_track_uri', 'episode_name',
       'episode_show_name', 'spotify_episode_uri', 'reason_start',
       'reason_end', 'shuffle', 'skipped', 'offline', 'offline_timestamp',
       'incognito_mode'],
      dtype='object')

In [7]:
uncleaned_extended[uncleaned_extended["offline"] == True]

ts    username  \
344   2014-08-11T04:00:58Z  allymarie7   
3136  2016-10-27T22:31:16Z  allymarie7   
3206  2016-11-03T20:08:14Z  allymarie7   
3570  2016-12-02T01:08:40Z  allymarie7   
7801  2018-01-10T03:12:59Z  allymarie7   
7803  2018-01-10T03:12:59Z  allymarie7   
7804  2018-01-10T03:12:59Z  allymarie7   
7805  2018-01-10T03:12:59Z  allymarie7   
7806  2018-01-10T03:12:59Z  allymarie7   
7808  2018-01-10T03:12:59Z  allymarie7   
8057  2018-01-26T20:12:53Z  allymarie7   
8058  2018-01-26T20:17:24Z  allymarie7   

                                            platform  ms_played conn_country  \
344   Android-tablet OS 4.4.4 API 19 (asus, Nexus 7)          0           US   
3136                           iOS 9.3.5 (iPhone8,4)     255093           US   
3206                           iOS 9.3.5 (iPhone8,4)     212592           US   
3570                           iOS 9.3.5 (iPhone8,4)     213520           US   
7801                            iOS 10.2 (iPhone8,4)     215987           US   
7803                            iOS 10.2 (iPhone8,4)     223260           US   
7804                            iOS 10.2 (iPhone8,4)     284840           US   
7805                            iOS 10.2 (iPhone8,4)     317302           US   
7806                            iOS 10.2 (iPhone8,4)     252651           US   
7808                            iOS 10.2 (iPhone8,4)     342209           US   
8057                            iOS 10.2 (iPhone8,4)      59153           US   
8058                            iOS 10.2 (iPhone8,4)     292093           US   

     ip_addr_decrypted user_agent_decrypted  \
344       172.56.11.59              unknown   
3136     98.222.243.22              unknown   
3206              None                 None   
3570              None                 None   
7801              None                 None   
7803              None                 None   
7804              None                 None   
7805              None                 None   
7806              None                 None   
7808              None                 None   
8057              None                 None   
8058              None                 None   

                    master_metadata_track_name  \
344                                 Pass Me By   
3136                               Hey Brother   
3206                        Can't Feel My Face   
3570                        Can't Feel My Face   
7801  Princess Mononoke - Instrumental Version   
7803              Ai Wa Hana Kimi Wa Sono Tane   
7804                  Kaze-No Tani-No Naushika   
7805                          Sekaino Yakusoku   
7806                        My Neighbor Totoro   
7808                        My Neighbor Totoro   
8057                        Sick Of Being Told   
8058          Spirited Away - One Summer's Day   

     master_metadata_album_artist_name  \
344                                 R5   
3136                            Avicii   
3206                        The Weeknd   
3570                        The Weeknd   
7801                      Relax α Wave   
7803                      Relax α Wave   
7804                      Relax α Wave   
7805                      Relax α Wave   
7806                      Relax α Wave   
7808                      Relax α Wave   
8057                  Grace VanderWaal   
8058                      Joe Hisaishi   

                       master_metadata_album_album_name  ... episode_name  \
344                                              Louder  ...         None   
3136                                               True  ...         None   
3206                          Beauty Behind The Madness  ...         None   
3570                          Beauty Behind The Madness  ...         None   
7801  Deep Sleep Music - The Best of Studio Ghibli: ...  ...         None   
7803  Deep Sleep Music - The Best of Studio Ghibli, ...  ...         None   
7804  Deep Sleep Music - The Best of Studio Ghibli: ...  ...         None   
7805  Deep

In [8]:
uncleaned_extended_artists = uncleaned_extended[["master_metadata_album_artist_name", "master_metadata_album_album_name", "master_metadata_track_name",  "spotify_track_uri", "skipped"]]

In [9]:
uncleaned_extended_artists.sort_values(by = ["master_metadata_album_album_name", "master_metadata_album_artist_name"]).head(50)

master_metadata_album_artist_name   master_metadata_album_album_name  \
6680                   Childish Gambino                 "Awaken, My Love!"   
7405                   Childish Gambino                 "Awaken, My Love!"   
9845                   Childish Gambino                 "Awaken, My Love!"   
12365                 The Neighbourhood  #000000 & #FFFFFF (No DJ Version)   
2752                    Various Artists                     #DayaTeenParty   
12675                          Dutchboy                         #FREESAUCE   
12743                        Rajitheone                          #TourLife   
897                              Swifty                     #swifty vol. 3   
1028                             Swifty                     #swifty vol. 3   
1047                             Swifty                     #swifty vol. 3   
1239                             Swifty                     #swifty vol. 3   
1259                             Swifty                     #swifty vol. 3   
1849                             Swifty                     #swifty vol. 3   
9145                             Jesse®                                  &   
11734                            Jesse®                                  &   
9167                      Billie Eilish         &burn (with Vince Staples)   
10031                     Billie Eilish         &burn (with Vince Staples)   
10050                     Billie Eilish         &burn (with Vince Staples)   
10071                     Billie Eilish         &burn (with Vince Staples)   
10087                     Billie Eilish         &burn (with Vince Staples)   
11281                     Billie Eilish         &burn (with Vince Staples)   
11282                     Billie Eilish         &burn (with Vince Staples)   
11433                     Billie Eilish         &burn (with Vince Staples)   
11834                     Billie Eilish         &burn (with Vince Staples)   
11874                     Billie Eilish         &burn (with Vince Staples)   
11875                     Billie Eilish         &burn (with Vince Staples)   
12745                     Billie Eilish         &burn (with Vince Staples)   
13459                     Billie Eilish         &burn (with Vince Staples)   
9736                         Ed Sheeran                                  +   
9781                         Ed Sheeran                                  +   
12931                        Ed Sheeran                                  +   
1271                        The Beatles                                  1   
11343                       The Beatles                                  1   
11348                       The Beatles                                  1   
11356                       The Beatles                                  1   
11359                       The Beatles                                  1   
11365                       The Beatles                                  1   
11369                       The Beatles                                  1   
11379                       The Beatles                                  1   
11380                       The Beatles                                  1   
11644                       The Beatles                                  1   
11647                       The Beatles                                  1   
11648                       The Beatles                                  1   
11662                       The Beatles                                  1   
11664                       The Beatles                                  1   
11669                       The Beatles                                  1   
11782                       The Beatles                                  1   
11796                       The Beatles                                  1   
11814                       The Beatles                                  1   
11841                       The Beatles                                  1   

                       mast

Create id column to join two dataframes

In [10]:
uncleaned_extended.loc[:, "id"] = uncleaned_extended.loc[:, "spotify_track_uri"].str[14:]

In [11]:
uncleaned_extended.head(10)

ts    username  \
0  2014-06-09T17:53:33Z  allymarie7   
1  2014-06-09T17:53:35Z  allymarie7   
2  2014-06-09T17:53:38Z  allymarie7   
3  2014-06-09T17:53:51Z  allymarie7   
4  2014-06-09T17:57:13Z  allymarie7   
5  2014-06-09T17:58:29Z  allymarie7   
6  2014-06-09T17:58:34Z  allymarie7   
7  2014-06-09T17:58:36Z  allymarie7   
8  2014-06-09T18:02:35Z  allymarie7   
9  2014-06-09T18:07:05Z  allymarie7   

                                         platform  ms_played conn_country  \
0  Android-tablet OS 4.4.2 API 19 (asus, Nexus 7)      11749           US   
1  Android-tablet OS 4.4.2 API 19 (asus, Nexus 7)       2554           US   
2  Android-tablet OS 4.4.2 API 19 (asus, Nexus 7)       3715           US   
3  Android-tablet OS 4.4.2 API 19 (asus, Nexus 7)      14485           US   
4  Android-tablet OS 4.4.2 API 19 (asus, Nexus 7)     202496           US   
5  Android-tablet OS 4.4.2 API 19 (asus, Nexus 7)      74907           US   
6  Android-tablet OS 4.4.2 API 19 (asus, Nexus 7)       5108           US   
7  Android-tablet OS 4.4.2 API 19 (asus, Nexus 7)       3065           US   
8  Android-tablet OS 4.4.2 API 19 (asus, Nexus 7)     239894           US   
9  Android-tablet OS 4.4.2 API 19 (asus, Nexus 7)     269560           US   

  ip_addr_decrypted user_agent_decrypted  \
0      50.165.42.23              unknown   
1      50.165.42.23              unknown   
2      50.165.42.23              unknown   
3      50.165.42.23              unknown   
4      50.165.42.23              unknown   
5      50.165.42.23              unknown   
6      50.165.42.23              unknown   
7      50.165.42.23              unknown   
8      50.165.42.23              unknown   
9      50.165.42.23              unknown   

                        master_metadata_track_name  \
0                                Beauty And A Beat   
1                             Timber (feat. Ke$ha)   
2  A Little Party Never Killed Nobody (All We Got)   
3                                          Fix You   
4                             She Looks So Perfect   
5                      Drunk in Love (feat. Jay-Z)   
6                  Waves - Robin Schulz Radio Edit   
7                    Rather Be (feat. Jess Glynne)   
8                                    Love Me Again   
9                                        All of Me   

  master_metadata_album_artist_name  \
0                     Justin Bieber   
1                           Pitbull   
2                            Fergie   
3                          Coldplay   
4               5 Seconds of Summer   
5                           Beyoncé   
6                         Mr. Probz   
7                      Clean Bandit   
8                       John Newman   
9                       John Legend   

                  master_metadata_album_album_name  ... episode_show_name  \
0                                 Believe Acoustic  ...              None   
1        Global Warming: Meltdown (Deluxe Version)  ...              None   
2  Music From Baz Luhrmann's Film The Great Gatsby  ...              None   
3                                              X&Y  ...              None   
4                             She Looks So Perfect  ...              None   
5                                          BEYONCÉ  ...              None   
6                                            Waves  ...              None   
7                    Rather Be (feat. Jess Glynne)  ...              None   
8                                          Tribute  ...              None   
9            Love In The Future (Expanded Edition)  ...              None   

  spotify_episode_uri reason_start reason_end shuffle skipped  offline  \
0                None      appload     fwdbtn   False     1.0    False   
1                None       fwdbtn     fwdbtn    True     1.0    False   
2                None       fwdbtn     fwdbtn    True     1.0    False   
3                None       fwdbtn    appload    True     1.0    False   
4              

## SONG DATA FROM KAGGLE

In [12]:
song_data = pd.read_csv("/Users/allisonkramp/Desktop/spring24/spotify project!/data.csv")

In [13]:
song_data[song_data["id"] == "4pbG9SUmWIvsROVLF0zF9s"]

acousticness          artists  danceability  duration_ms  energy  \
85954         0.386  ['The Beatles']          0.49       145747   0.715   

       explicit                      id  instrumentalness  key  liveness  \
85954         0  4pbG9SUmWIvsROVLF0zF9s               0.0    7     0.311   

       loudness  mode                                        name  popularity  \
85954    -5.549     1  I Want To Hold Your Hand - Remastered 2015          74   

      release_date  speechiness    tempo  valence  year  
85954   2000-11-13       0.0476  130.726    0.866  2000

In [14]:
song_data

acousticness                                   artists  danceability  \
0             0.9950                       ['Carl Woitschach']         0.708   
1             0.9940  ['Robert Schumann', 'Vladimir Horowitz']         0.379   
2             0.6040                   ['Seweryn Goszczyński']         0.749   
3             0.9950                      ['Francisco Canaro']         0.781   
4             0.9900  ['Frédéric Chopin', 'Vladimir Horowitz']         0.210   
...              ...                                       ...           ...   
169904        0.1730                    ['DripReport', 'Tyga']         0.875   
169905        0.0167        ['Leon Bridges', 'Terrace Martin']         0.719   
169906        0.5380                     ['Kygo', 'Oh Wonder']         0.514   
169907        0.0714             ['Cash Cash', 'Andy Grammer']         0.646   
169908        0.1090                        ['Ingrid Andress']         0.512   

        duration_ms  energy  explicit                      id  \
0            158648  0.1950         0  6KbQ3uYMLKb5jDxLF7wYDD   
1            282133  0.0135         0  6KuQTIu1KoTTkLXKrwlLPV   
2            104300  0.2200         0  6L63VW0PibdM1HDSBoqnoM   
3            180760  0.1300         0  6M94FkXd15sOAOQYRnWPN8   
4            687733  0.2040         0  6N6tiFZ9vLTSOIxkj8qKrd   
...             ...     ...       ...                     ...   
169904       163800  0.4430         1  4KppkflX7I3vJQk7urOJaS   
169905       167468  0.3850         0  1ehhGlTvjtHo2e4xJFB0SZ   
169906       180700  0.5390         0  52eycxprLhK3lPcRLbQiVk   
169907       167308  0.7610         0  3wYOGJYD31sLRmBgCvWxa4   
169908       214787  0.4280         0  60RFlt48hm0l4Fu0JoccOl   

        instrumentalness  key  liveness  loudness  mode  \
0               0.563000   10    0.1510   -12.428     1   
1               0.901000    8    0.0763   -28.454     1   
2               0.000000    5    0.1190   -19.924     0   
3               0.887000    1    0.1110   -14.734     0   
4               0.908000   11    0.0980   -16.829     1   
...                  ...  ...       ...       ...   ...   
169904          0.000032    1    0.0891    -7.461     1   
169905          0.031300    8    0.1110   -10.907     1   
169906          0.002330    7    0.1080    -9.332     1   
169907          0.000000    1    0.2220    -2.557     1   
169908          0.000000    0    0.1050    -7.387     1   

                                                 name  popularity  \
0                         Singende Bataillone 1. Teil           0   
1            Fantasiestücke, Op. 111: Più tosto lento           0   
2                      Chapter 1.18 - Zamek kaniowski           0   
3       Bebamos Juntos - Instrumental (Remasterizado)           0   
4         Polonaise-Fantaisie in A-Flat Major, Op. 61           1   
...                                               ...         ...   
169904                  Skechers (feat. Tyga) - Remix          75   
169905                 Sweeter (feat. Terrace Martin)          64   
169906                               How Would I Know          70   
169907                                    I Found You          70   
169908                          More Hearts Than Mine          65   

       release_date  speechiness    tempo  valence  year  
0              1928       0.0506  118.469   0.7790  1928  
1              1928       0.0462   83.972   0.0767  1928  
2              1928       0.9290  107.177   0.8800  1928  
3        1928-09-25       0.0926  108.003   0.7200  1928  
4              1928       0.0424   62.149   0.0693  1928  
...             ...          ...      ...      ...   ...  
169904   2020-05-15       0.1430  100.012   0.3060  2020  
169905   2020-06-08       0.0403  128.000   0.2700  2020  
169906   2020-05-29       0.1050  123.700   0.1530  2020  
169907   2020-02-28       0.0385  129.916   0.4720  2020  
169908   2020-03-27       0.0271   80.588   0.3660  2020  

[169909 rows x 19 c

In [15]:
full_df = song_data.merge(uncleaned_extended, how="inner", on="id", suffixes = ('_songdata', '_extend'))

In [16]:
full_df.columns

Index(['acousticness', 'artists', 'danceability', 'duration_ms', 'energy',
       'explicit', 'id', 'instrumentalness', 'key', 'liveness', 'loudness',
       'mode', 'name', 'popularity', 'release_date', 'speechiness', 'tempo',
       'valence', 'year', 'ts', 'username', 'platform', 'ms_played',
       'conn_country', 'ip_addr_decrypted', 'user_agent_decrypted',
       'master_metadata_track_name', 'master_metadata_album_artist_name',
       'master_metadata_album_album_name', 'spotify_track_uri', 'episode_name',
       'episode_show_name', 'spotify_episode_uri', 'reason_start',
       'reason_end', 'shuffle', 'skipped', 'offline', 'offline_timestamp',
       'incognito_mode'],
      dtype='object')

In [17]:
full_df.head(10)

acousticness                                   artists  danceability  \
0       0.99400  ['Johann Sebastian Bach', 'Glenn Gould']         0.405   
1       0.39300                          ['Mariah Carey']         0.532   
2       0.00871                        ['Hannah Montana']         0.629   
3       0.86400                           ['The Strokes']         0.736   
4       0.86400                           ['The Strokes']         0.736   
5       0.86400                           ['The Strokes']         0.736   
6       0.81800                            ['Bruno Mars']         0.643   
7       0.81800                            ['Bruno Mars']         0.643   
8       0.54000                          ['Darlene Love']         0.353   
9       0.54000                          ['Darlene Love']         0.353   

   duration_ms  energy  explicit                      id  instrumentalness  \
0        94707   0.277         0  2OodraMlK1o898KizHf4MI          0.814000   
1       277093   0.456         0  3GONDOCwJprD1lZZ32JbuJ          0.000002   
2       194973   0.901         0  7jg5JfizOZSEipnTR6d4Nt          0.000000   
3       182027   0.297         0  6z5mh42xSM54RncKZMM8hS          0.000015   
4       182027   0.297         0  6z5mh42xSM54RncKZMM8hS          0.000015   
5       182027   0.297         0  6z5mh42xSM54RncKZMM8hS          0.000015   
6       249535   0.394         0  7Cj2tT0m2SBeIO1PsGfk8A          0.000000   
7       249535   0.394         0  7Cj2tT0m2SBeIO1PsGfk8A          0.000000   
8       165651   0.703         0  2Px4PrbPBMeWB1silbjI5H          0.064100   
9       165651   0.703         0  2Px4PrbPBMeWB1silbjI5H          0.064100   

   key  liveness  ...  episode_name  episode_show_name spotify_episode_uri  \
0    0    0.1140  ...          None               None                None   
1    9    0.1380  ...          None               None                None   
2    4    0.0573  ...          None               None                None   
3    2    0.0609  ...          None               None                None   
4    2    0.0609  ...          None               None                None   
5    2    0.0609  ...          None               None                None   
6    2    0.0888  ...          None               None                None   
7    2    0.0888  ...          None               None                None   
8    1    0.9100  ...          None               None                None   
9    1    0.9100  ...          None               None                None   

   reason_start                    reason_end  shuffle  skipped  offline  \
0      clickrow                          None    False      NaN    False   
1        fwdbtn                          None     True      NaN    False   
2     trackdone                          None    False      NaN    False   
3     trackdone                     trackdone     True      NaN    False   
4     trackdone                     trackdone     True      NaN    False   
5     trackdone                     trackdone     True      NaN    False   
6     trackdone                          None    False      NaN    False   
7       appload                          None    False      NaN    False   
8     trackdone  unexpected-exit-while-paused     True      NaN    False   
9     trackdone                       endplay     True      NaN    False   

   offline_timestamp incognito_mode  
0      1516678605702          False  
1                  0          False  
2                  0          False  
3      1561338094381          False  
4      1561396802527          False  
5      1561432090437          False  
6      1491182049711          False  
7      1491183899520          False  
8      1561247914048          False  
9      1561256991838          False  

[10 rows x 40 columns]

Create column for percent of song played

In [18]:
full_df['percent_played'] = (full_df['ms_played'] / full_df['duration_ms']) * 100

In [19]:
full_df[['ms_played', 'duration_ms', 'percent_played']]

ms_played  duration_ms  percent_played
0         94706        94707       99.998944
1          4226       277093        1.525120
2        194973       194973      100.000000
3        182026       182027       99.999451
4        182026       182027       99.999451
...         ...          ...             ...
8546     174320       174320      100.000000
8547      49433       174320       28.357618
8548     174320       174320      100.000000
8549     270519       270520       99.999630
8550       6757       215227        3.139476

[8551 rows x 3 columns]

In [20]:
full_df["ts"]

0       2018-01-23T03:38:20Z
1       2016-11-09T00:34:49Z
2       2016-12-05T00:07:06Z
3       2019-06-24T01:04:37Z
4       2019-06-24T17:23:05Z
                ...         
8546    2019-09-25T04:24:49Z
8547    2019-09-25T17:17:11Z
8548    2019-10-01T03:44:10Z
8549    2019-06-24T18:50:50Z
8550    2019-05-31T01:59:22Z
Name: ts, Length: 8551, dtype: object

#make sure song name matches and song artist matches

In [21]:
full_df[["name", "artists", "master_metadata_album_artist_name",  "master_metadata_track_name", "master_metadata_album_album_name", "id", "spotify_track_uri"]]

name  \
0     Prelude and Fugue No. 2 in C Minor, BWV 847: F...   
1                                          If It's Over   
2     Who Said - From "Hannah Montana"/Soundtrack Ve...   
3                                          Call Me Back   
4                                          Call Me Back   
...                                                 ...   
8546                                           The King   
8547                                           The King   
8548                                           The King   
8549  WHEN I WAS OLDER - Music Inspired By The Film ...   
8550                                         Right Back   

                                       artists  \
0     ['Johann Sebastian Bach', 'Glenn Gould']   
1                             ['Mariah Carey']   
2                           ['Hannah Montana']   
3                              ['The Strokes']   
4                              ['The Strokes']   
...                                        ...   
8546                            ['Conan Gray']   
8547                            ['Conan Gray']   
8548                            ['Conan Gray']   
8549                         ['Billie Eilish']   
8550                                ['Khalid']   

     master_metadata_album_artist_name  \
0                Johann Sebastian Bach   
1                         Mariah Carey   
2                       Hannah Montana   
3                          The Strokes   
4                          The Strokes   
...                                ...   
8546                        Conan Gray   
8547                        Conan Gray   
8548                        Conan Gray   
8549                     Billie Eilish   
8550                            Khalid   

                             master_metadata_track_name  \
0     Prelude and Fugue No. 2 in C Minor, BWV 847: F...   
1                                          If It's Over   
2     Who Said - From "Hannah Montana"/Soundtrack Ve...   
3                                          Call Me Back   
4                                          Call Me Back   
...                                                 ...   
8546                                           The King   
8547                                           The King   
8548                                           The King   
8549  WHEN I WAS OLDER - Music Inspired By The Film ...   
8550                                         Right Back   

                       master_metadata_album_album_name  \
0     Bach: The Well-Tempered Clavier, Book I, Prelu...   
1                                              Emotions   
2                                        Hannah Montana   
3                                                Angles   
4                                                Angles   
...                                                 ...   
8546                                           The King   
8547                                           The King   
8548                                           The King   
8549                                   WHEN I WAS OLDER   
8550                                        Free Spirit   

                          id                     spotify_track_uri  
0     2OodraMlK1o898KizHf4MI  spotify:track:2OodraMlK1o898KizHf4MI  
1     3GONDOCwJprD1lZZ32JbuJ  spotify:track:3GONDOCwJprD1lZZ32JbuJ  
2     7jg5JfizOZSEipnTR6d4Nt  spotify:track:7jg5JfizOZSEipnTR6d4Nt  
3     6z5mh42xSM54RncKZMM8hS  spotify:track:6z5mh42xSM54RncKZMM8hS  
4     6z5mh42xSM54RncKZMM8hS  spotify:track:6z5mh42xSM54RncKZMM8hS  
...                      ...                                   ...  
8546  54GASuUdLbhhdyECNUp3Pb  spotify:track:54GASuUdLbhhdyECNUp3Pb  
8547  54GASuUdLbhhdyECNUp3Pb  spotify:track:54GASuUdLbhhdyECNUp3Pb  
8548  54GASuUdLbhhdyECNUp3Pb  spotify:track:54GASuUdLbhhdyECNUp3Pb  
8549  7tGEAA1f8MydT7eVbbO9Zy  spotify:track:7tGEAA1f8MydT7eVbbO9Zy  
8550  5A1w94uzc1kO1Zhe8WWxC3  spotify:track:5A1w94uzc1kO1Zh

In [22]:
full_df.columns

Index(['acousticness', 'artists', 'danceability', 'duration_ms', 'energy',
       'explicit', 'id', 'instrumentalness', 'key', 'liveness', 'loudness',
       'mode', 'name', 'popularity', 'release_date', 'speechiness', 'tempo',
       'valence', 'year', 'ts', 'username', 'platform', 'ms_played',
       'conn_country', 'ip_addr_decrypted', 'user_agent_decrypted',
       'master_metadata_track_name', 'master_metadata_album_artist_name',
       'master_metadata_album_album_name', 'spotify_track_uri', 'episode_name',
       'episode_show_name', 'spotify_episode_uri', 'reason_start',
       'reason_end', 'shuffle', 'skipped', 'offline', 'offline_timestamp',
       'incognito_mode', 'percent_played'],
      dtype='object')

## PREPARE DATA SET FOR NN

idt i want the song title in it? cuz what would that tell, but artist name and albumn name might help cuz ppl might like songs from the same artist / album
maybe remove: 'conn_country', 'ip_addr_decrypted', 'user_agent_decrypted', episode_name, episode_show_name, 'spotify_episode_uri', 'offline_timestamp', 'incognito_mode', 'platform', 'release date' (might put that one back in), 

need to remove all podcasts

shuffle - if shuffle mode was used while listening, skipped - whether you skipped to next song

In [23]:
full_df["episode_name"].isnull().sum() #no podcasts in dataset

8551

In [24]:
all_columns = full_df[['acousticness', 'danceability', 'duration_ms', 'energy',
       'explicit', 'instrumentalness', 'key', 'liveness', 'loudness',
       'mode', 'popularity', 'speechiness', 'tempo',
       'valence', 'year', 'percent_played',  'master_metadata_track_name', 'master_metadata_album_artist_name',
       'master_metadata_album_album_name', 'reason_start',
       'reason_end', 'shuffle', 'skipped', 'offline']]

In [25]:
all_columns.head(10)

acousticness  danceability  duration_ms  energy  explicit  \
0       0.99400         0.405        94707   0.277         0   
1       0.39300         0.532       277093   0.456         0   
2       0.00871         0.629       194973   0.901         0   
3       0.86400         0.736       182027   0.297         0   
4       0.86400         0.736       182027   0.297         0   
5       0.86400         0.736       182027   0.297         0   
6       0.81800         0.643       249535   0.394         0   
7       0.81800         0.643       249535   0.394         0   
8       0.54000         0.353       165651   0.703         0   
9       0.54000         0.353       165651   0.703         0   

   instrumentalness  key  liveness  loudness  mode  ...  year  percent_played  \
0          0.814000    0    0.1140   -21.501     0  ...  1963       99.998944   
1          0.000002    9    0.1380    -9.673     1  ...  1991        1.525120   
2          0.000000    4    0.0573    -3.357     1  ...  2006      100.000000   
3          0.000015    2    0.0609    -9.648     1  ...  2011       99.999451   
4          0.000015    2    0.0609    -9.648     1  ...  2011       99.999451   
5          0.000015    2    0.0609    -9.648     1  ...  2011       99.999451   
6          0.000000    2    0.0888    -6.305     0  ...  2011        1.577735   
7          0.000000    2    0.0888    -6.305     0  ...  2011       98.422265   
8          0.064100    1    0.9100    -5.847     0  ...  2011       95.759156   
9          0.064100    1    0.9100    -5.847     0  ...  2011        4.482919   

                          master_metadata_track_name  \
0  Prelude and Fugue No. 2 in C Minor, BWV 847: F...   
1                                       If It's Over   
2  Who Said - From "Hannah Montana"/Soundtrack Ve...   
3                                       Call Me Back   
4                                       Call Me Back   
5                                       Call Me Back   
6                                 Grenade - Acoustic   
7                                 Grenade - Acoustic   
8              (Today I Met) The Boy I'm Gonna Marry   
9              (Today I Met) The Boy I'm Gonna Marry   

   master_metadata_album_artist_name  \
0              Johann Sebastian Bach   
1                       Mariah Carey   
2                     Hannah Montana   
3                        The Strokes   
4                        The Strokes   
5                        The Strokes   
6                         Bruno Mars   
7                         Bruno Mars   
8                       Darlene Love   
9                       Darlene Love   

                    master_metadata_album_album_name  reason_start  \
0  Bach: The Well-Tempered Clavier, Book I, Prelu...      clickrow   
1                                           Emotions        fwdbtn   
2                                     Hannah Montana     trackdone   
3                                             Angles     trackdone   
4                                             Angles     trackdone   
5                                             Angles     trackdone   
6                               The Grenade Sessions     trackdone   
7                               The Grenade Sessions       appload   
8   The Sound Of Love: The Very Best Of Darlene Love     trackdone   
9   The Sound Of Love: The Very Best Of Darlene Love     trackdone   

                     reason_end shuffle skipped offline  
0                          None   False     NaN   False  
1                          None    True     NaN   False  
2                          None   False     NaN   False  
3                     trackdone    True     NaN   False  
4                     trackdone    True     NaN   False  
5                     trackdone    True     NaN   False  
6                          None   False     NaN   False  
7                          None   False     NaN   False  
8  unexpected-exit-while-paused    True     NaN   False

In [26]:
reason_ended = all_columns["reason_end"].unique()
reason_ended

array([None, 'trackdone', 'unexpected-exit-while-paused', 'endplay',
       'backbtn', 'appload', 'fwdbtn', 'unexpected-exit', 'logout',
       'unknown', 'remote'], dtype=object)

I think I should remove all the ones that arent : trackdone, endplay, backbtn, fwdbtn

In [27]:
prep_df = all_columns[all_columns["reason_end"].isin(['trackdone', 'endplay', 'backbtn', 'fwdbtn'])]
prep_df.head(10)

acousticness  danceability  duration_ms  energy  explicit  \
3       0.864000         0.736       182027   0.297         0   
4       0.864000         0.736       182027   0.297         0   
5       0.864000         0.736       182027   0.297         0   
9       0.540000         0.353       165651   0.703         0   
10      0.000203         0.507       246770   0.951         0   
11      0.000203         0.507       246770   0.951         0   
12      0.000203         0.507       246770   0.951         0   
13      0.000203         0.507       246770   0.951         0   
14      0.000203         0.507       246770   0.951         0   
15      0.000203         0.507       246770   0.951         0   

    instrumentalness  key  liveness  loudness  mode  ...  year  \
3           0.000015    2    0.0609    -9.648     1  ...  2011   
4           0.000015    2    0.0609    -9.648     1  ...  2011   
5           0.000015    2    0.0609    -9.648     1  ...  2011   
9           0.064100    1    0.9100    -5.847     0  ...  2011   
10          0.478000    1    0.4170    -3.710     1  ...  2012   
11          0.478000    1    0.4170    -3.710     1  ...  2012   
12          0.478000    1    0.4170    -3.710     1  ...  2012   
13          0.478000    1    0.4170    -3.710     1  ...  2012   
14          0.478000    1    0.4170    -3.710     1  ...  2012   
15          0.478000    1    0.4170    -3.710     1  ...  2012   

    percent_played             master_metadata_track_name  \
3        99.999451                           Call Me Back   
4        99.999451                           Call Me Back   
5        99.999451                           Call Me Back   
9         4.482919  (Today I Met) The Boy I'm Gonna Marry   
10       20.714025                              Centipede   
11        6.416501                              Centipede   
12        1.354703                              Centipede   
13       23.994813                              Centipede   
14        1.748592                              Centipede   
15        1.166673                              Centipede   

    master_metadata_album_artist_name  \
3                         The Strokes   
4                         The Strokes   
5                         The Strokes   
9                        Darlene Love   
10                        Knife Party   
11                        Knife Party   
12                        Knife Party   
13                        Knife Party   
14                        Knife Party   
15                        Knife Party   

                    master_metadata_album_album_name  reason_start reason_end  \
3                                             Angles     trackdone  trackdone   
4                                             Angles     trackdone  trackdone   
5                                             Angles     trackdone  trackdone   
9   The Sound Of Love: The Very Best Of Darlene Love     trackdone    endplay   
10                                       Rage Valley     trackdone    backbtn   
11                                       Rage Valley     trackdone    backbtn   
12                                       Rage Valley     trackdone    backbtn   
13                                       Rage Valley     trackdone    backbtn   
14                                       Rage Valley     trackdone    backbtn   
15                                       Rage Valley     trackdone    backbtn   

   shuffle skipped offline  
3     True     NaN   False  
4     True     NaN   False  
5     True     NaN   False  
9     True     NaN   False  
10   False     1.0   False  
11   False     1.0   False  
12   False     1.0   False  
13   False     1.0   False  
14   False     1.0   False  
15   False     1.0   False  

[10 rows x 24 columns]

### Encode whether song is liked by the reason it ended

In [28]:
def likeable(row):
    if row["reason_end"] in ["trackdone"]:
        return 1
    else:
        return 0

In [29]:
prep_df.loc[:, 'liked'] = prep_df.apply(likeable, axis=1)

/var/folders/12/nb5wd0692kqdjk5ls7v3ww980000gn/T/ipykernel_20119/3510299924.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  prep_df.loc[:, 'liked'] = prep_df.apply(likeable, axis=1)


### Encode reason_end - Have to do One Hot encoding bc no order in categories

In [30]:
prep_df = prep_df.reset_index(drop=True)
prep_df

acousticness  danceability  duration_ms  energy  explicit  \
0         0.864000         0.736       182027   0.297         0   
1         0.864000         0.736       182027   0.297         0   
2         0.864000         0.736       182027   0.297         0   
3         0.540000         0.353       165651   0.703         0   
4         0.000203         0.507       246770   0.951         0   
...            ...           ...          ...     ...       ...   
5156      0.064300         0.643       174320   0.716         0   
5157      0.064300         0.643       174320   0.716         0   
5158      0.064300         0.643       174320   0.716         0   
5159      0.853000         0.696       270520   0.332         0   
5160      0.064900         0.788       215227   0.427         0   

      instrumentalness  key  liveness  loudness  mode  ...  percent_played  \
0             0.000015    2    0.0609    -9.648     1  ...       99.999451   
1             0.000015    2    0.0609    -9.648     1  ...       99.999451   
2             0.000015    2    0.0609    -9.648     1  ...       99.999451   
3             0.064100    1    0.9100    -5.847     0  ...        4.482919   
4             0.478000    1    0.4170    -3.710     1  ...       20.714025   
...                ...  ...       ...       ...   ...  ...             ...   
5156          0.000000    2    0.1400    -6.733     1  ...      100.000000   
5157          0.000000    2    0.1400    -6.733     1  ...       28.357618   
5158          0.000000    2    0.1400    -6.733     1  ...      100.000000   
5159          0.433000    6    0.0848   -12.834     1  ...       99.999630   
5160          0.000036    5    0.1710    -9.918     0  ...        3.139476   

                             master_metadata_track_name  \
0                                          Call Me Back   
1                                          Call Me Back   
2                                          Call Me Back   
3                 (Today I Met) The Boy I'm Gonna Marry   
4                                             Centipede   
...                                                 ...   
5156                                           The King   
5157                                           The King   
5158                                           The King   
5159  WHEN I WAS OLDER - Music Inspired By The Film ...   
5160                                         Right Back   

      master_metadata_album_artist_name  \
0                           The Strokes   
1                           The Strokes   
2                           The Strokes   
3                          Darlene Love   
4                           Knife Party   
...                                 ...   
5156                         Conan Gray   
5157                         Conan Gray   
5158                         Conan Gray   
5159                      Billie Eilish   
5160                             Khalid   

                      master_metadata_album_album_name  reason_start  \
0                                               Angles     trackdone   
1                                               Angles     trackdone   
2                                               Angles     trackdone   
3     The Sound Of Love: The Very Best Of Darlene Love     trackdone   
4                                          Rage Valley     trackdone   
...                                                ...           ...   
5156                                          The King     trackdone   
5157                                          The King     trackdone   
5158                                          The King     trackdone   
5159                                  WHEN I WAS OLDER     trackdone   
5160                                       Free Spirit        fwdbtn   

      reason_end shuffle skipped offline liked  
0      trackdone    True     NaN   False     1  
1      trackdone    True     NaN   False     1  
2      trackdone    True  

In [31]:
# importing one hot encoder 
from sklearn.preprocessing import OneHotEncoder

In [32]:

# creating one hot encoder object 
onehotencoder = OneHotEncoder()
#reshape the 1-D country array to 2-D as fit_transform expects 2-D and finally fit the object 
X = onehotencoder.fit_transform(prep_df.reason_end.values.reshape(-1,1)).toarray()
print(onehotencoder.categories_)
print(X)
print(X.shape)

[array(['backbtn', 'endplay', 'fwdbtn', 'trackdone'], dtype=object)]
[[0. 0. 0. 1.]
 [0. 0. 0. 1.]
 [0. 0. 0. 1.]
 ...
 [0. 0. 0. 1.]
 [0. 0. 0. 1.]
 [0. 0. 1. 0.]]
(5161, 4)


In [33]:
#To add this back into the original dataframe 
dfOneHot = pd.DataFrame(X, columns = ['backbtn_end', 'endplay_end', 'fwdbtn_end', 'trackdone_end']) 
dfOneHot

backbtn_end  endplay_end  fwdbtn_end  trackdone_end
0             0.0          0.0         0.0            1.0
1             0.0          0.0         0.0            1.0
2             0.0          0.0         0.0            1.0
3             0.0          1.0         0.0            0.0
4             1.0          0.0         0.0            0.0
...           ...          ...         ...            ...
5156          0.0          0.0         0.0            1.0
5157          0.0          0.0         1.0            0.0
5158          0.0          0.0         0.0            1.0
5159          0.0          0.0         0.0            1.0
5160          0.0          0.0         1.0            0.0

[5161 rows x 4 columns]

In [34]:
reason_end_enc = pd.concat([prep_df, dfOneHot], axis=1)
reason_end_enc

acousticness  danceability  duration_ms  energy  explicit  \
0         0.864000         0.736       182027   0.297         0   
1         0.864000         0.736       182027   0.297         0   
2         0.864000         0.736       182027   0.297         0   
3         0.540000         0.353       165651   0.703         0   
4         0.000203         0.507       246770   0.951         0   
...            ...           ...          ...     ...       ...   
5156      0.064300         0.643       174320   0.716         0   
5157      0.064300         0.643       174320   0.716         0   
5158      0.064300         0.643       174320   0.716         0   
5159      0.853000         0.696       270520   0.332         0   
5160      0.064900         0.788       215227   0.427         0   

      instrumentalness  key  liveness  loudness  mode  ...  reason_start  \
0             0.000015    2    0.0609    -9.648     1  ...     trackdone   
1             0.000015    2    0.0609    -9.648     1  ...     trackdone   
2             0.000015    2    0.0609    -9.648     1  ...     trackdone   
3             0.064100    1    0.9100    -5.847     0  ...     trackdone   
4             0.478000    1    0.4170    -3.710     1  ...     trackdone   
...                ...  ...       ...       ...   ...  ...           ...   
5156          0.000000    2    0.1400    -6.733     1  ...     trackdone   
5157          0.000000    2    0.1400    -6.733     1  ...     trackdone   
5158          0.000000    2    0.1400    -6.733     1  ...     trackdone   
5159          0.433000    6    0.0848   -12.834     1  ...     trackdone   
5160          0.000036    5    0.1710    -9.918     0  ...        fwdbtn   

      reason_end  shuffle  skipped  offline  liked backbtn_end endplay_end  \
0      trackdone     True      NaN    False      1         0.0         0.0   
1      trackdone     True      NaN    False      1         0.0         0.0   
2      trackdone     True      NaN    False      1         0.0         0.0   
3        endplay     True      NaN    False      0         0.0         1.0   
4        backbtn    False      1.0    False      0         1.0         0.0   
...          ...      ...      ...      ...    ...         ...         ...   
5156   trackdone    False      NaN    False      1         0.0         0.0   
5157      fwdbtn    False      NaN    False      0         0.0         0.0   
5158   trackdone    False      NaN    False      1         0.0         0.0   
5159   trackdone     True      NaN    False      1         0.0         0.0   
5160      fwdbtn     True      NaN    False      0         0.0         0.0   

     fwdbtn_end trackdone_end  
0           0.0           1.0  
1           0.0           1.0  
2           0.0           1.0  
3           0.0           0.0  
4           0.0           0.0  
...         ...           ...  
5156        0.0           1.0  
5157        1.0           0.0  
5158        0.0           1.0  
5159        0.0           1.0  
5160        1.0           0.0  

[5161 rows x 29 columns]

### Reason start Encoding

In [35]:
# creating one hot encoder object 
onehotencoder2 = OneHotEncoder()
#reshape the 1-D country array to 2-D as fit_transform expects 2-D and finally fit the object 
X2 = onehotencoder2.fit_transform(prep_df.reason_start.values.reshape(-1,1)).toarray()
print(onehotencoder2.categories_)
print(X2)
print(X2.shape)

[array(['appload', 'backbtn', 'clickrow', 'fwdbtn', 'playbtn', 'remote',
       'trackdone', 'trackerror'], dtype=object)]
[[0. 0. 0. ... 0. 1. 0.]
 [0. 0. 0. ... 0. 1. 0.]
 [0. 0. 0. ... 0. 1. 0.]
 ...
 [0. 0. 0. ... 0. 1. 0.]
 [0. 0. 0. ... 0. 1. 0.]
 [0. 0. 0. ... 0. 0. 0.]]
(5161, 8)


In [36]:
#To add this back into the original dataframe 
dfOneHot2 = pd.DataFrame(X2, columns = ['appload_start', 'backbtn_start', 'clickrow_start', 'fwdbtn_start', 'playbtn_start', 'remote_start', 'trackdone_start', 'trackerror_start']) 
dfOneHot2

appload_start  backbtn_start  clickrow_start  fwdbtn_start  \
0               0.0            0.0             0.0           0.0   
1               0.0            0.0             0.0           0.0   
2               0.0            0.0             0.0           0.0   
3               0.0            0.0             0.0           0.0   
4               0.0            0.0             0.0           0.0   
...             ...            ...             ...           ...   
5156            0.0            0.0             0.0           0.0   
5157            0.0            0.0             0.0           0.0   
5158            0.0            0.0             0.0           0.0   
5159            0.0            0.0             0.0           0.0   
5160            0.0            0.0             0.0           1.0   

      playbtn_start  remote_start  trackdone_start  trackerror_start  
0               0.0           0.0              1.0               0.0  
1               0.0           0.0              1.0               0.0  
2               0.0           0.0              1.0               0.0  
3               0.0           0.0              1.0               0.0  
4               0.0           0.0              1.0               0.0  
...             ...           ...              ...               ...  
5156            0.0           0.0              1.0               0.0  
5157            0.0           0.0              1.0               0.0  
5158            0.0           0.0              1.0               0.0  
5159            0.0           0.0              1.0               0.0  
5160            0.0           0.0              0.0               0.0  

[5161 rows x 8 columns]

In [37]:
final_df = pd.concat([reason_end_enc, dfOneHot2], axis=1)
final_df

acousticness  danceability  duration_ms  energy  explicit  \
0         0.864000         0.736       182027   0.297         0   
1         0.864000         0.736       182027   0.297         0   
2         0.864000         0.736       182027   0.297         0   
3         0.540000         0.353       165651   0.703         0   
4         0.000203         0.507       246770   0.951         0   
...            ...           ...          ...     ...       ...   
5156      0.064300         0.643       174320   0.716         0   
5157      0.064300         0.643       174320   0.716         0   
5158      0.064300         0.643       174320   0.716         0   
5159      0.853000         0.696       270520   0.332         0   
5160      0.064900         0.788       215227   0.427         0   

      instrumentalness  key  liveness  loudness  mode  ...  fwdbtn_end  \
0             0.000015    2    0.0609    -9.648     1  ...         0.0   
1             0.000015    2    0.0609    -9.648     1  ...         0.0   
2             0.000015    2    0.0609    -9.648     1  ...         0.0   
3             0.064100    1    0.9100    -5.847     0  ...         0.0   
4             0.478000    1    0.4170    -3.710     1  ...         0.0   
...                ...  ...       ...       ...   ...  ...         ...   
5156          0.000000    2    0.1400    -6.733     1  ...         0.0   
5157          0.000000    2    0.1400    -6.733     1  ...         1.0   
5158          0.000000    2    0.1400    -6.733     1  ...         0.0   
5159          0.433000    6    0.0848   -12.834     1  ...         0.0   
5160          0.000036    5    0.1710    -9.918     0  ...         1.0   

      trackdone_end  appload_start  backbtn_start  clickrow_start  \
0               1.0            0.0            0.0             0.0   
1               1.0            0.0            0.0             0.0   
2               1.0            0.0            0.0             0.0   
3               0.0            0.0            0.0             0.0   
4               0.0            0.0            0.0             0.0   
...             ...            ...            ...             ...   
5156            1.0            0.0            0.0             0.0   
5157            0.0            0.0            0.0             0.0   
5158            1.0            0.0            0.0             0.0   
5159            1.0            0.0            0.0             0.0   
5160            0.0            0.0            0.0             0.0   

      fwdbtn_start playbtn_start remote_start trackdone_start trackerror_start  
0              0.0           0.0          0.0             1.0              0.0  
1              0.0           0.0          0.0             1.0              0.0  
2              0.0           0.0          0.0             1.0              0.0  
3              0.0           0.0          0.0             1.0              0.0  
4              0.0           0.0          0.0             1.0              0.0  
...            ...           ...          ...             ...              ...  
5156           0.0           0.0          0.0             1.0              0.0  
5157           0.0           0.0          0.0             1.0              0.0  
5158           0.0           0.0          0.0             1.0              0.0  
5159           0.0           0.0          0.0             1.0              0.0  
5160           1.0           0.0          0.0             0.0              0.0  

[5161 rows x 37 columns]

In [38]:
final_df.iloc[20:30, 18:]

master_metadata_album_album_name reason_start reason_end  shuffle  skipped  \
20              5 Seconds Of Summer    trackdone  trackdone    False      0.0   
21              5 Seconds Of Summer    trackdone  trackdone    False      0.0   
22              5 Seconds Of Summer      backbtn    backbtn    False      1.0   
23              5 Seconds Of Summer      appload  trackdone    False      0.0   
24              5 Seconds Of Summer    trackdone  trackdone    False      0.0   
25              5 Seconds Of Summer    trackdone  trackdone    False      0.0   
26              5 Seconds Of Summer    trackdone  trackdone    False      0.0   
27              5 Seconds Of Summer    trackdone  trackdone    False      0.0   
28              5 Seconds Of Summer    trackdone  trackdone     True      NaN   
29              5 Seconds Of Summer    trackdone    endplay    False      NaN   

    offline  liked  backbtn_end  endplay_end  fwdbtn_end  trackdone_end  \
20    False      1          0.0          0.0         0.0            1.0   
21    False      1          0.0          0.0         0.0            1.0   
22    False      0          1.0          0.0         0.0            0.0   
23    False      1          0.0          0.0         0.0            1.0   
24    False      1          0.0          0.0         0.0            1.0   
25    False      1          0.0          0.0         0.0            1.0   
26    False      1          0.0          0.0         0.0            1.0   
27    False      1          0.0          0.0         0.0            1.0   
28    False      1          0.0          0.0         0.0            1.0   
29    False      0          0.0          1.0         0.0            0.0   

    appload_start  backbtn_start  clickrow_start  fwdbtn_start  playbtn_start  \
20            0.0            0.0             0.0           0.0            0.0   
21            0.0            0.0             0.0           0.0            0.0   
22            0.0            1.0             0.0           0.0            0.0   
23            1.0            0.0             0.0           0.0            0.0   
24            0.0            0.0             0.0           0.0            0.0   
25            0.0            0.0             0.0           0.0            0.0   
26            0.0            0.0             0.0           0.0            0.0   
27            0.0            0.0             0.0           0.0            0.0   
28            0.0            0.0             0.0           0.0            0.0   
29            0.0            0.0             0.0           0.0            0.0   

    remote_start  trackdone_start  trackerror_start  
20           0.0              1.0               0.0  
21           0.0              1.0               0.0  
22           0.0              0.0               0.0  
23           0.0              0.0               0.0  
24           0.0              1.0               0.0  
25           0.0              1.0               0.0  
26           0.0              1.0               0.0  
27           0.0              1.0               0.0  
28           0.0              1.0               0.0  
29           0.0              1.0               0.0

In [39]:
#droping the country column 
final_df = final_df.drop(['reason_start', 'reason_end'], axis=1) 
#printing to verify 
print(final_df.head())

   acousticness  danceability  duration_ms  energy  explicit  \
0      0.864000         0.736       182027   0.297         0   
1      0.864000         0.736       182027   0.297         0   
2      0.864000         0.736       182027   0.297         0   
3      0.540000         0.353       165651   0.703         0   
4      0.000203         0.507       246770   0.951         0   

   instrumentalness  key  liveness  loudness  mode  ...  fwdbtn_end  \
0          0.000015    2    0.0609    -9.648     1  ...         0.0   
1          0.000015    2    0.0609    -9.648     1  ...         0.0   
2          0.000015    2    0.0609    -9.648     1  ...         0.0   
3          0.064100    1    0.9100    -5.847     0  ...         0.0   
4          0.478000    1    0.4170    -3.710     1  ...         0.0   

   trackdone_end  appload_start  backbtn_start  clickrow_start  fwdbtn_start  \
0            1.0            0.0            0.0             0.0           0.0   
1            1.0            

Encode Shuffle and Offline to be 0 -1 not TF

In [40]:
final_df.loc[:, "shuffle"] = final_df["shuffle"].astype(int)
final_df.loc[:, "offline"] = final_df["offline"].astype(int)

/var/folders/12/nb5wd0692kqdjk5ls7v3ww980000gn/T/ipykernel_20119/1586733319.py:1: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '[1 1 1 ... 0 1 1]' has dtype incompatible with bool, please explicitly cast to a compatible dtype first.
  final_df.loc[:, "shuffle"] = final_df["shuffle"].astype(int)
/var/folders/12/nb5wd0692kqdjk5ls7v3ww980000gn/T/ipykernel_20119/1586733319.py:2: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '[0 0 0 ... 0 0 0]' has dtype incompatible with bool, please explicitly cast to a compatible dtype first.
  final_df.loc[:, "offline"] = final_df["offline"].astype(int)


In [41]:
final_df.iloc[0:10, 18:27]

master_metadata_album_album_name  shuffle  skipped  \
0                                            Angles        1      NaN   
1                                            Angles        1      NaN   
2                                            Angles        1      NaN   
3  The Sound Of Love: The Very Best Of Darlene Love        1      NaN   
4                                       Rage Valley        0      1.0   
5                                       Rage Valley        0      1.0   
6                                       Rage Valley        0      1.0   
7                                       Rage Valley        0      1.0   
8                                       Rage Valley        0      1.0   
9                                       Rage Valley        0      1.0   

   offline  liked  backbtn_end  endplay_end  fwdbtn_end  trackdone_end  
0        0      1          0.0          0.0         0.0            1.0  
1        0      1          0.0          0.0         0.0            1.0  
2        0      1          0.0          0.0         0.0            1.0  
3        0      0          0.0          1.0         0.0            0.0  
4        0      0          1.0          0.0         0.0            0.0  
5        0      0          1.0          0.0         0.0            0.0  
6        0      0          1.0          0.0         0.0            0.0  
7        0      0          1.0          0.0         0.0            0.0  
8        0      0          1.0          0.0         0.0            0.0  
9        0      0          1.0          0.0         0.0            0.0

In [42]:
final_df.head()

acousticness  danceability  duration_ms  energy  explicit  \
0      0.864000         0.736       182027   0.297         0   
1      0.864000         0.736       182027   0.297         0   
2      0.864000         0.736       182027   0.297         0   
3      0.540000         0.353       165651   0.703         0   
4      0.000203         0.507       246770   0.951         0   

   instrumentalness  key  liveness  loudness  mode  ...  fwdbtn_end  \
0          0.000015    2    0.0609    -9.648     1  ...         0.0   
1          0.000015    2    0.0609    -9.648     1  ...         0.0   
2          0.000015    2    0.0609    -9.648     1  ...         0.0   
3          0.064100    1    0.9100    -5.847     0  ...         0.0   
4          0.478000    1    0.4170    -3.710     1  ...         0.0   

   trackdone_end  appload_start  backbtn_start  clickrow_start  fwdbtn_start  \
0            1.0            0.0            0.0             0.0           0.0   
1            1.0            0.0            0.0             0.0           0.0   
2            1.0            0.0            0.0             0.0           0.0   
3            0.0            0.0            0.0             0.0           0.0   
4            0.0            0.0            0.0             0.0           0.0   

  playbtn_start remote_start trackdone_start  trackerror_start  
0           0.0          0.0             1.0               0.0  
1           0.0          0.0             1.0               0.0  
2           0.0          0.0             1.0               0.0  
3           0.0          0.0             1.0               0.0  
4           0.0          0.0             1.0               0.0  

[5 rows x 35 columns]

### ENCODE TRAKCNAME, ALBUM NAME, ARTIST NAME

okay here the thing, i'm dropping album name and track name. track name is just unnecesssary. Album name I wanted to keep so that the algorithm could learn which albumns are liked best. However, tracks in an album should have a correlation with their tempo, acousticness, ect. so album should not be needed.

LEARNING WITH COUNTS?? instead of Onehotencoding

In [43]:
((3+23)/(492+35)) *100

4.933586337760911

In [44]:
len(final_df["master_metadata_album_album_name"].unique())

777

In [45]:
len(final_df["master_metadata_album_artist_name"].unique())

492

## CLEAN THE DATASET SOME MORE 

In [46]:
prep_df.isnull().sum()

acousticness                            0
danceability                            0
duration_ms                             0
energy                                  0
explicit                                0
instrumentalness                        0
key                                     0
liveness                                0
loudness                                0
mode                                    0
popularity                              0
speechiness                             0
tempo                                   0
valence                                 0
year                                    0
percent_played                          0
master_metadata_track_name              0
master_metadata_album_artist_name       0
master_metadata_album_album_name        0
reason_start                            0
reason_end                              0
shuffle                                 0
skipped                              4826
offline                           

In [47]:
prep_df.loc[:, "skipped"] = prep_df["skipped"].fillna(0)

In [48]:
prep_df.head(10)

acousticness  danceability  duration_ms  energy  explicit  \
0      0.864000         0.736       182027   0.297         0   
1      0.864000         0.736       182027   0.297         0   
2      0.864000         0.736       182027   0.297         0   
3      0.540000         0.353       165651   0.703         0   
4      0.000203         0.507       246770   0.951         0   
5      0.000203         0.507       246770   0.951         0   
6      0.000203         0.507       246770   0.951         0   
7      0.000203         0.507       246770   0.951         0   
8      0.000203         0.507       246770   0.951         0   
9      0.000203         0.507       246770   0.951         0   

   instrumentalness  key  liveness  loudness  mode  ...  percent_played  \
0          0.000015    2    0.0609    -9.648     1  ...       99.999451   
1          0.000015    2    0.0609    -9.648     1  ...       99.999451   
2          0.000015    2    0.0609    -9.648     1  ...       99.999451   
3          0.064100    1    0.9100    -5.847     0  ...        4.482919   
4          0.478000    1    0.4170    -3.710     1  ...       20.714025   
5          0.478000    1    0.4170    -3.710     1  ...        6.416501   
6          0.478000    1    0.4170    -3.710     1  ...        1.354703   
7          0.478000    1    0.4170    -3.710     1  ...       23.994813   
8          0.478000    1    0.4170    -3.710     1  ...        1.748592   
9          0.478000    1    0.4170    -3.710     1  ...        1.166673   

              master_metadata_track_name  master_metadata_album_artist_name  \
0                           Call Me Back                        The Strokes   
1                           Call Me Back                        The Strokes   
2                           Call Me Back                        The Strokes   
3  (Today I Met) The Boy I'm Gonna Marry                       Darlene Love   
4                              Centipede                        Knife Party   
5                              Centipede                        Knife Party   
6                              Centipede                        Knife Party   
7                              Centipede                        Knife Party   
8                              Centipede                        Knife Party   
9                              Centipede                        Knife Party   

                   master_metadata_album_album_name  reason_start  reason_end  \
0                                            Angles     trackdone   trackdone   
1                                            Angles     trackdone   trackdone   
2                                            Angles     trackdone   trackdone   
3  The Sound Of Love: The Very Best Of Darlene Love     trackdone     endplay   
4                                       Rage Valley     trackdone     backbtn   
5                                       Rage Valley     trackdone     backbtn   
6                                       Rage Valley     trackdone     backbtn   
7                                       Rage Valley     trackdone     backbtn   
8                                       Rage Valley     trackdone     backbtn   
9                                       Rage Valley     trackdone     backbtn   

  shuffle skipped offline liked  
0    True     0.0   False     1  
1    True     0.0   False     1  
2    True     0.0   False     1  
3    True     0.0   False     0  
4   False     1.0   False     0  
5   False     1.0   False     0  
6   False     1.0   False     0  
7   False     1.0   False     0  
8   False     1.0   False     0  
9   False     1.0   False     0  

[10 rows x 25 columns]

In [49]:
prep_df.iloc[0:10, 9:17]

mode  popularity  speechiness    tempo  valence  year  percent_played  \
0     1          54       0.0349  125.851   0.1860  2011       99.999451   
1     1          54       0.0349  125.851   0.1860  2011       99.999451   
2     1          54       0.0349  125.851   0.1860  2011       99.999451   
3     0          54       0.0583  100.076   0.6080  2011        4.482919   
4     1          54       0.0654  140.077   0.0628  2012       20.714025   
5     1          54       0.0654  140.077   0.0628  2012        6.416501   
6     1          54       0.0654  140.077   0.0628  2012        1.354703   
7     1          54       0.0654  140.077   0.0628  2012       23.994813   
8     1          54       0.0654  140.077   0.0628  2012        1.748592   
9     1          54       0.0654  140.077   0.0628  2012        1.166673   

              master_metadata_track_name  
0                           Call Me Back  
1                           Call Me Back  
2                           Call Me Back  
3  (Today I Met) The Boy I'm Gonna Marry  
4                              Centipede  
5                              Centipede  
6                              Centipede  
7                              Centipede  
8                              Centipede  
9                              Centipede

### check value ranges for each column

In [50]:
prep_df["acousticness"].round(1).unique()

array([0.9, 0.5, 0. , 0.1, 0.2, 0.4, 0.6, 0.8, 1. , 0.7, 0.3])

In [51]:
prep_df["danceability"].round(1).unique()

array([0.7, 0.4, 0.5, 0.3, 0.1, 0.8, 0.9, 0.6, 0.2, 1. ])

In [52]:
prep_df["duration_ms"].round(-5).unique()

array([200000, 300000, 100000,      0, 400000, 500000])

In [53]:
prep_df["energy"].round(1).unique()

array([0.3, 0.7, 1. , 0.6, 0.2, 0.9, 0.8, 0.4, 0.5, 0. , 0.1])

In [54]:
prep_df["explicit"].round(1).unique()

array([0, 1])

In [55]:
prep_df["instrumentalness"].round(1).unique()

array([0. , 0.1, 0.5, 0.9, 0.2, 0.4, 0.3, 0.8, 0.7, 0.6, 1. ])

In [56]:
prep_df["key"].unique()

array([ 2,  1,  5,  0,  6,  9,  4,  7, 10,  8, 11,  3])

In [57]:
prep_df["liveness"].round(1).unique()

array([0.1, 0.9, 0.4, 0.2, 0.3, 0.6, 0.5, 0.7, 0. , 0.8])

In [58]:
prep_df["loudness"].round(1).unique()

array([ -9.6,  -5.8,  -3.7,  -9. , -17.7,  -4.8,  -2.7,  -6.6,  -4.4,
        -4.5,  -5.9,  -8.4,  -4.2,  -5.6,  -6.4,  -5.7, -10.2,  -4.6,
       -10.5,  -5. ,  -6.5,  -4.9,  -8.1, -10.3,  -8.2,  -6.1,  -7.4,
        -4.1,  -5.5, -12.6,  -6. , -16.8, -14. ,  -7.8,  -9.4,  -6.8,
        -5.1, -17.9,  -7.2,  -9.5,  -5.4,  -8.6,  -7.5, -11.1,  -7.1,
       -36.3,  -4. , -10.1,  -2. , -11.3, -12.8,  -7.9,  -7.6, -12. ,
        -8. ,  -5.2, -11.2,  -9.2,  -3.3, -10. ,  -5.3,  -9.8, -36.6,
        -3.2,  -4.3, -11. ,  -2.8,  -7.7,  -3.1, -10.7,  -9.3,  -6.2,
       -15.9, -11.9, -27.5,  -3.4,  -4.7, -14.5,  -3.5, -17.5,  -8.5,
        -7. , -12.3,  -8.8,  -6.9,  -2.2, -16.9,  -8.3,  -9.7,  -3.8,
       -11.6,  -2.6,  -8.9, -20.9,  -6.3, -10.6, -14.7, -11.5,  -8.7,
       -22. ,  -6.7, -20.4,  -2.9, -14.4,  -3.9,  -3.6,  -9.1, -14.8,
       -14.6, -20.5, -21.3, -14.1, -14.3, -15.5, -13.9, -12.4, -23.6,
       -16.5, -10.4,  -9.9, -13.6, -10.8, -20.8, -11.8, -11.4, -16.7,
       -12.5, -13.2,

In [59]:
prep_df["mode"].round(1).unique()

array([1, 0])

In [60]:
prep_df["popularity"].unique()

array([54, 53, 55, 52, 61, 59, 57, 60, 63, 64, 62, 67, 65, 68, 66, 50, 69,
       70, 28, 48, 58, 51, 44, 56, 49, 71, 18, 41, 46, 47, 40, 78, 74, 73,
       45, 79, 77, 76, 72, 80, 83, 82, 75, 81, 85, 91, 92, 84, 87, 86, 89,
       93, 38, 42, 39])

In [61]:
prep_df["speechiness"].round(1).unique()

array([0. , 0.1, 0.4, 0.3, 0.2, 0.9, 0.6, 0.5])

In [62]:
prep_df["tempo"].round().unique()

array([126., 100., 140., 180.,  75., 150., 144., 128.,  96., 104., 156.,
       127., 164., 105., 113., 119., 125., 102., 116., 114.,  78., 190.,
        92.,  87.,  98., 133., 134.,  85.,  77.,  76., 163., 151., 135.,
       170.,  89., 139., 130., 157., 142., 112., 129.,  83.,  97.,  81.,
        39., 174., 146.,  79., 137., 117.,  70., 118., 115., 160.,  74.,
        84., 120., 124., 101., 109., 108., 110., 143., 162., 182., 152.,
       111., 132., 147.,  82.,  90.,  80., 123., 138., 131., 153., 175.,
       159., 122., 202., 141.,  95., 136., 121., 169., 158.,  93., 107.,
       106., 103., 171., 148., 200., 168., 154.,  73.,  72., 176., 149.,
       145., 155.,  88.,  99.,  94., 166., 184., 179., 177.,  65.,  63.,
       165.,  71., 178., 167., 161., 206., 186.,  68.,  91.,  86., 173.,
       192., 204., 201., 172.,  62., 208., 187.])

In [63]:
prep_df["year"].round().unique()

array([2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019, 2010, 1970,
       1999, 1963, 2001, 2007, 2009, 2004, 2008, 1933, 1955, 1956, 1957,
       1958, 1959, 1961, 1962, 1964, 1965, 1966, 1967, 1968, 1969, 1971,
       1972, 1973, 1976, 1977, 1979, 1980, 1983, 1984, 1985, 1986, 1992,
       1993, 1995, 2000, 2002, 2003, 2006, 1990, 1994, 1997, 1998, 1981,
       2005])

In [64]:
prep_df["percent_played"].round().unique()

array([100.,   4.,  21.,   6.,   1.,  24.,   2.,  99.,  14.,   7.,  12.,
        15.,   0.,  95.,   3.,  63.,  37.,   5.,  29.,  83.,  28.,  39.,
        58.,  18.,  20.,  49.,  23.,  25.,   9.,  89.,  16.,  11.,  84.,
        22.,  67., 138.,  31.,  91.,  38.,  34.,  46.,  56.,  43.,  50.,
        26.,  75.,  94.,  64.,  30.,  48.,  62.,  19.,  96.,  86.,  35.,
         8., 158.,  17.,  10.,  71.,  85.,  98.,  54.,  13.,  27.,  81.,
        47.,  82.,  74.,  69.,  66.,  32., 124., 107.,  87.,  79.,  44.,
        40.,  59.,  77.,  41., 101.,  88.,  36.,  53.,  97., 159.,  68.,
        61., 198.,  42.,  90.,  65.,  70.,  73.,  55., 154.,  76.,  60.,
        57.,  80., 113.,  78., 123.,  33.,  92.,  52.,  45., 146., 139.,
        72.])

IT SAYS REASON_START AND REASON_END ARE array([0], dtype=object) AND THATS WHY ITS GIVING ME AN ERROR

In [65]:
prep_df["reason_start"].unique()

array(['trackdone', 'backbtn', 'appload', 'clickrow', 'fwdbtn', 'playbtn',
       'trackerror', 'remote'], dtype=object)

In [66]:
prep_df["reason_end"].unique()

array(['trackdone', 'endplay', 'backbtn', 'fwdbtn'], dtype=object)

In [67]:
prep_df["shuffle"].unique()

array([ True, False])

In [68]:
prep_df["offline"].unique()

array([False,  True])

In [69]:
prep_df["skipped"].unique()

array([0., 1.])

### done checking values in columns

In [70]:
# prep_df.reset_index(drop=True)

In [71]:
5161 * .7

3612.7

In [72]:
5161 - 3612

1549

train: 3612 test: 1549

## begin neural net

later I want to change the like score based on a range of how many times I listened to the song

In [73]:
#using pytorch cuz its not static like tensorflow, tensorflow also takes longer train but less memory
import os
import torch
from torch import nn
from torch.utils.data import DataLoader
from torchvision import datasets, transforms
import torch.optim as optim
import torchvision
import torchvision.transforms as transforms
import torchvision.datasets as datasets

In [74]:
print(torch.__version__)

2.2.0


In [75]:
# CREATE Y SPLIT - FOR LABELING
final_df_y = prep_df["liked"]
len(final_df_y)

5161

In [76]:
# CREATE X SPLIT - WITHOUT LABELS
final_df_x = prep_df.iloc[:,0:24]
num_features = len(final_df_x.columns)
final_df_x.columns

Index(['acousticness', 'danceability', 'duration_ms', 'energy', 'explicit',
       'instrumentalness', 'key', 'liveness', 'loudness', 'mode', 'popularity',
       'speechiness', 'tempo', 'valence', 'year', 'percent_played',
       'master_metadata_track_name', 'master_metadata_album_artist_name',
       'master_metadata_album_album_name', 'reason_start', 'reason_end',
       'shuffle', 'skipped', 'offline'],
      dtype='object')

In [77]:
# TURN TO NUMPY TO REMOVE COLUMN NAMES
numpy_x = prep_df.values
numpy_x.shape

(5161, 25)

In [78]:
# TURN Y TO NUMPY
numpy_y = final_df_y.values
numpy_y.shape

(5161,)

splitting typical 2/3 train , 1/3 test for rn - might change later

Should I do dimensionality reduction?

- if there is overfitting / takes too long I will
however, not too many features might be unecessary
- but do I want feature engineering cuz idk which attributes are unneccessary

In [79]:
train, test = torch.utils.data.random_split(numpy_x, [.7, .3])

In [80]:
train.dataset

array([[0.8640000000000001, 0.736, 182027, ..., 0.0, False, 1],
       [0.8640000000000001, 0.736, 182027, ..., 0.0, False, 1],
       [0.8640000000000001, 0.736, 182027, ..., 0.0, False, 1],
       ...,
       [0.0643, 0.643, 174320, ..., 0.0, False, 1],
       [0.853, 0.696, 270520, ..., 0.0, False, 1],
       [0.0649, 0.7879999999999999, 215227, ..., 0.0, False, 0]],
      dtype=object)

batch size of 100: train 242 rows at a time

In [81]:
batch_size = 242

In [82]:
input_dim = 24
hidden_dim = 10
output_dim = 2

In [83]:
# Convert DataFrame to PyTorch tensor
features_tensor = torch.tensor((prep_df.iloc[:, 0:24]).values, dtype=torch.float64)
target_tensor = torch.tensor(prep_df.iloc[:, 25:].values, dtype=torch.int64)

TypeError: can't convert np.ndarray of type numpy.object_. The only supported types are: float64, float32, float16, complex64, complex128, int64, int32, int16, int8, uint8, and bool.

: 

: 

In [ ]:
train_loader = torch.utils.data.DataLoader(dataset=train, 
                                           batch_size=batch_size, 
                                           shuffle=True)

test_loader = torch.utils.data.DataLoader(dataset=test, 
                                          batch_size=batch_size, 
                                          shuffle=False)


: 

In [ ]:
class_label = train_loader [:, -1] # for last column
class_label

: 

: 

: 

: 

: 

: 

Define the architecture of the feedforward neural network

In [ ]:
class FeedforwardNN(nn.Module):
    def __init__(self, input_dim, hidden_dim, output_dim, cuda=False):
        super(FeedforwardNN, self).__init__()
        
        # Linear function
        self.model = nn.Sequential(
            nn.Linear(input_dim, hidden_dim),
            nn.Sigmoid(),
            nn.Linear(hidden_dim, output_dim)
        ) 

    def forward(self, x):
        return self.model(x)

: 

In [ ]:
# Instantiate the model
model = FeedforwardNN(input_dim, hidden_dim, output_dim, ).to('cpu')
print(f"model: {model}")

# Define loss function and optimizer
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model.parameters(), lr=0.001)

: 

In [ ]:
for epoch in range(10):
    for i, (vals, labels) in enumerate(train_loader):
        pass

: 

: 

: 

: 

### should I create a tally of the amount of times the song is listened to?

: 

: 

: 

: 

: 

: 

: 

: 

: 

## IDK WHATS GOING ON DOWN HERE

In [ ]:
my_song_data.iloc[0:20,0:20]

: 

In [ ]:
uncleaned_extended["spotify_track_uri_cleaned"] = ""

: 

In [ ]:
uncleaned_extended["spotify_track_uri_cleaned"] = uncleaned_extended["spotify_track_uri"].str.rindex(":")

: 

In [ ]:
uncleaned_extended

: 

In [ ]:
result = uncleaned_extended.loc[:,["spotify_track_uri_cleaned"]] == 13
result

: 

In [ ]:
uncleaned_extended.iloc[1487] #all the songs that say false are just none and need to get rid of those songs

: 

In [ ]:
result.loc[result["spotify_track_uri_cleaned"] == False]

: 